File: myExpenses/src/main/java/org/totschnig/myexpenses/activity/CsvImportActivity.java
package org.totschnig.myexpenses.activity;

import android.os.Bundle;

import com.google.android.material.snackbar.Snackbar;

import org.apache.commons.csv.CSVRecord;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.dialog.ConfirmationDialogFragment;
import org.totschnig.myexpenses.export.qif.QifDateFormat;
import org.totschnig.myexpenses.fragment.CsvImportDataFragment;
import org.totschnig.myexpenses.fragment.CsvImportParseFragment;
import org.totschnig.myexpenses.model.AccountType;
import org.totschnig.myexpenses.model.ContribFeature;
import org.totschnig.myexpenses.model.CurrencyUnit;
import org.totschnig.myexpenses.task.TaskExecutionFragment;
import org.totschnig.myexpenses.util.Result;

import java.util.ArrayList;

import androidx.appcompat.app.ActionBar;
import icepick.State;


public class CsvImportActivity extends TabbedActivity implements
    ConfirmationDialogFragment.ConfirmationDialogListener {

  @State
  boolean mDataReady = false;
  @State
  boolean mUsageRecorded = false;

  private void setmDataReady(boolean mDataReady) {
    this.mDataReady = mDataReady;
    mSectionsPagerAdapter.notifyDataSetChanged();
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    final ActionBar actionBar = getSupportActionBar();
    actionBar.setTitle(getString(R.string.pref_import_title, "CSV"));
  }

  @Override
  protected void setupTabs() {
    //we only add the first tab, the second one once data has been parsed
    addTab(0);
    if (mDataReady) {
      addTab(1);
    }
  }

  private void addTab(int index) {
    switch (index) {
      case 0:
        mSectionsPagerAdapter.addFragment(CsvImportParseFragment.newInstance(), getString(
            R.string.menu_parse));
        break;
      case 1:
        mSectionsPagerAdapter.addFragment(CsvImportDataFragment.newInstance(), getString(
            R.string.csv_import_preview));
        break;
    }
  }

  @Override
  public void onPositive(Bundle args) {
    switch (args.getInt(ConfirmationDialogFragment.KEY_COMMAND_POSITIVE)) {
      case R.id.SET_HEADER_COMMAND:
        CsvImportDataFragment df = (CsvImportDataFragment) getSupportFragmentManager().findFragmentByTag(
            mSectionsPagerAdapter.getFragmentName(1));
        df.setHeader();
    }
  }

  @Override
  public void onNegative(Bundle args) {

  }

  @Override
  public void onDismissOrCancel(Bundle args) {

  }

  @Override
  public void onPostExecute(int taskId, Object result) {
    super.onPostExecute(taskId, result);
    switch (taskId) {
      case TaskExecutionFragment.TASK_CSV_PARSE:
        if (result != null) {
          ArrayList<CSVRecord> data = (ArrayList<CSVRecord>) result;
          if (!data.isEmpty()) {
            if (!mDataReady) {
              addTab(1);
              setmDataReady(true);
            }
            CsvImportDataFragment df = (CsvImportDataFragment) getSupportFragmentManager().findFragmentByTag(
                mSectionsPagerAdapter.getFragmentName(1));
            if (df != null) {
              df.setData(data);
              mViewPager.setCurrentItem(1);
            }
            break;
          }
        }
        showSnackbar(R.string.parse_error_no_data_found, Snackbar.LENGTH_LONG);
        break;
      case TaskExecutionFragment.TASK_CSV_IMPORT:
        Result r = (Result) result;
        String msg;
        if (r.isSuccess()) {
          if (!mUsageRecorded) {
            recordUsage(ContribFeature.CSV_IMPORT);
            mUsageRecorded = true;
          }
        }
        showSnackbar(r.print(this), Snackbar.LENGTH_LONG);
    }
  }

  @Override
  public void onProgressUpdate(Object progress) {
    if (progress instanceof String) {
      showSnackbar((String) progress, Snackbar.LENGTH_LONG);
    } else {
      super.onProgressUpdate(progress);
    }
  }

  public long getAccountId() {
    CsvImportParseFragment pf = getParseFragment();
    return pf.getAccountId();
  }

  private CsvImportParseFragment getParseFragment() {
    return (CsvImportParseFragment) getSupportFragmentManager().findFragmentByTag(
        mSectionsPagerAdapter.getFragmentName(0));
  }

  public CurrencyUnit getCurrency() {
    CsvImportParseFragment pf = getParseFragment();
    return currencyContext.get(pf.getCurrency());
  }

  public QifDateFormat getDateFormat() {
    CsvImportParseFragment pf = getParseFragment();
    return pf.getDateFormat();
  }

  public AccountType getAccountType() {
    CsvImportParseFragment pf = getParseFragment();
    return pf.getAccountType();
  }

}

File: myExpenses/src/main/java/org/totschnig/myexpenses/activity/CsvImportActivity.kt
package org.totschnig.myexpenses.activity

import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.widget.AdapterView
import androidx.activity.OnBackPressedCallback
import androidx.activity.viewModels
import androidx.viewpager2.widget.ViewPager2
import com.evernote.android.state.State
import org.apache.commons.csv.CSVRecord
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.db2.Repository
import org.totschnig.myexpenses.dialog.ConfirmationDialogFragment
import org.totschnig.myexpenses.dialog.ConfirmationDialogFragment.ConfirmationDialogListener
import org.totschnig.myexpenses.dialog.MessageDialogFragment
import org.totschnig.myexpenses.export.qif.QifDateFormat
import org.totschnig.myexpenses.fragment.CsvImportDataFragment
import org.totschnig.myexpenses.fragment.CsvImportParseFragment
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.model.AccountType
import org.totschnig.myexpenses.model.ContribFeature
import org.totschnig.myexpenses.model.ContribFeatureNotAvailableException
import org.totschnig.myexpenses.util.crashreporting.CrashHandler
import org.totschnig.myexpenses.viewmodel.AccountConfiguration
import org.totschnig.myexpenses.viewmodel.CsvImportViewModel
import org.totschnig.myexpenses.viewmodel.CsvImportViewModel.Companion.KEY_HEADER_LINE_POSITION
import java.io.FileNotFoundException
import javax.inject.Inject


class CsvImportActivity : TabbedActivity(), ConfirmationDialogListener {

    @State
    var mUsageRecorded = false

    @State
    var idle = true

    @Inject
    lateinit var repository: Repository

    private val csvImportViewModel: CsvImportViewModel by viewModels()

    private lateinit var onBackPressedCallback: OnBackPressedCallback

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar?.title = getString(R.string.pref_import_title, "CSV")
        with(injector) {
            inject(csvImportViewModel)
            inject(this@CsvImportActivity)
        }
        onBackPressedCallback = object : OnBackPressedCallback(false) {
            override fun handleOnBackPressed() {
                val currentItem = binding.viewPager.currentItem
                if (currentItem > 0) {
                    binding.viewPager.currentItem = currentItem - 1
                } else {
                    onBackPressedDispatcher.onBackPressed()

                }
            }
        }
        onBackPressedDispatcher.addCallback(this, onBackPressedCallback)
        binding.viewPager.registerOnPageChangeCallback(object :
            ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                onBackPressedCallback.isEnabled = position > 0
            }
        })
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val allowed = parseFragment?.isReady == true && idle
        menu.findItem(R.id.PARSE_COMMAND)?.isEnabled = allowed
        menu.findItem(R.id.IMPORT_COMMAND)?.isEnabled = allowed
        super.onPrepareOptionsMenu(menu)
        return true
    }

    override fun dispatchCommand(command: Int, tag: Any?) = when {
        super.dispatchCommand(command, tag) -> true
        command == R.id.CLOSE_COMMAND -> {
            finish()
            true
        }
        else -> false
    }

    override fun doHome() {
        if (binding.viewPager.currentItem == 1) {
            binding.viewPager.currentItem = 0
        } else super.doHome()
    }

    override fun createFragment(position: Int) = when(position) {
        0 -> CsvImportParseFragment.newInstance()
        1 -> CsvImportDataFragment.newInstance()
        else -> throw IllegalArgumentException()
    }

    override fun getItemCount() = 2

    override fun getTitle(position: Int) = when(position) {
        0 -> getString(R.string.menu_parse)
        1 -> getString(R.string.preview)
        else -> throw IllegalArgumentException()
    }


    override fun onPositive(args: Bundle, checked: Boolean) {
        super.onPositive(args, checked)
        if (args.getInt(ConfirmationDialogFragment.KEY_COMMAND_POSITIVE) == R.id.SET_HEADER_COMMAND) {
            dataFragment!!.setHeader(args.getInt(KEY_HEADER_LINE_POSITION))
        }
    }

    private fun showProgress() {
        showProgressSnackBar(getString(R.string.pref_import_title, "CSV"))
        idle = false
        invalidateOptionsMenu()
    }

    private fun hideProgress() {
        dismissSnackBar()
        idle = true
        invalidateOptionsMenu()
    }

    fun parseFile(uri: Uri, delimiter: Char, encoding: String) {
        showProgress()
        csvImportViewModel.withAccountColumn = accountId == 0L
        csvImportViewModel.parseFile(uri, delimiter, encoding).observe(this) { result ->
            hideProgress()
            result.onSuccess {
                binding.viewPager.currentItem = 1
            }.onFailure {
                showSnackBar(
                    when (it) {
                        is FileNotFoundException -> getString(
                            R.string.parse_error_file_not_found,
                            uri
                        )
                        else -> getString(R.string.parse_error_other_exception, it.message)
                    }
                )
            }
        }
    }

    fun importData(dataSet: List<CSVRecord>, columnToFieldMap: IntArray, discardedRows: Int) {
        accountId.takeIf { it != AdapterView.INVALID_ROW_ID }?.also { accountId ->
            showProgress()
            csvImportViewModel.importData(
                dataSet,
                columnToFieldMap,
                dateFormat,
                parseFragment!!.autoFillCategories,
                AccountConfiguration(accountId, currency, accountType),
                parseFragment!!.uri!!
            ).observe(this) { result ->
                hideProgress()
                result.onSuccess { resultList ->
                    if (!mUsageRecorded) {
                        recordUsage(ContribFeature.CSV_IMPORT)
                        mUsageRecorded = true
                    }
                    val msg = StringBuilder()
                    if (discardedRows > 0) {
                        msg.append(" ${getString(R.string.csv_import_records_discarded, discardedRows)}")
                    }
                    msg.append(resultList.joinToString(" ") {
                        "${getString(R.string.import_transactions_success, it.successCount, it.label)}."
                    })

                    showMessage(
                        msg,
                        neutral = MessageDialogFragment.nullButton(R.string.button_label_continue),
                        positive = MessageDialogFragment.Button(
                            R.string.button_label_close,
                            R.id.CLOSE_COMMAND,
                            null
                        )
                    )
                }.onFailure {
                    if (it !is ContribFeatureNotAvailableException) {
                        CrashHandler.report(it)
                    }
                    showSnackBar(it.message ?: it.javaClass.simpleName)
                }
            }
        } ?: kotlin.run {
            val exception = Exception("No account selected")
            CrashHandler.report(exception)
            showSnackBar(exception.message!!)
        }
    }

    private val dataFragment: CsvImportDataFragment?
        get() = supportFragmentManager.findFragmentByTag(
            mSectionsPagerAdapter.getFragmentName(1)
        ) as CsvImportDataFragment?

    private val parseFragment: CsvImportParseFragment?
        get() = supportFragmentManager.findFragmentByTag(
            mSectionsPagerAdapter.getFragmentName(0)
        ) as? CsvImportParseFragment

    val accountId: Long
        get() {
            return parseFragment!!.getSelectedAccountId()
        }

    val currency: String
        get() = parseFragment!!.getSelectedCurrency()

    val dateFormat: QifDateFormat
        get() {
            return parseFragment!!.dateFormat
        }

    val accountType: AccountType
        get() {
            return parseFragment!!.accountType
        }
}